# submitted by Tarang Ranpara (202011057)

In [ ]:
! pip install surprise

     |████████████████████████████████| 11.8MB 16.6MB/s 
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.1-cp37-cp37m-linux_x86_64.whl size=1617566 sha256=534b62006be7262f68ee39c165d870e05eabb25c97300c923dcb8d467b70ef03
  Stored in directory: /root/.cache/pip/wheels/78/9c/3d/41b419c9d2aff5b6e2b4c0fc8d25c538202834058f9ed110d0
Successfully built scikit-surprise


In [ ]:
# importing essential libs
from surprise import Dataset
from surprise import KNNBaseline
from surprise import accuracy
from surprise.model_selection import train_test_split
from surprise.model_selection import cross_validate

In [ ]:
# loading the data
data = Dataset.load_builtin(name='ml-100k')

Dataset ml-100k could not be found. Do you want to download it? [Y/n] Y
Trying to download dataset from http://files.grouplens.org/datasets/movielens/ml-100k.zip...
Done! Dataset ml-100k has been saved to /root/.surprise_data/ml-100k


In [ ]:
# splitting dataset
trainset, testset = train_test_split(data, test_size=0.2)

## User-based collaborative Filtering
- basic KNN algorithm with top 30 neighbours/users for prediction
- similarity measure : cosine similarity
- Stochastic Gradient Descent(SGD) for baseline estimation
- Cross-validating the approach on the training set
- using RMSE on the testing set for accuracy 

In [ ]:
sim_options = {"name": "cosine", "user-based": True}
algo_ub = KNNBaseline(k=40, sim_options=sim_options)

In [ ]:
# Run 5-fold cross-validation
cross_validate(algo_ub, data, measures=['RMSE'], cv=5, verbose=True)

Estimating biases using als...
Computing the cosine similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the cosine similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the cosine similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the cosine similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the cosine similarity matrix...
Done computing similarity matrix.
Evaluating RMSE of algorithm KNNBaseline on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9293  0.9358  0.9324  0.9417  0.9310  0.9340  0.0044  
Fit time          1.45    1.44    1.46    1.47    1.46    1.46    0.01    
Test time         4.55    4.52    4.52    4.60    4.45    4.53    0.05    


{'fit_time': (1.4478976726531982,
  1.4417569637298584,
  1.4565179347991943,
  1.474522590637207,
  1.4629449844360352),
 'test_rmse': array([0.92931013, 0.93575334, 0.93238   , 0.9416522 , 0.9310349 ]),
 'test_time': (4.553791284561157,
  4.518980264663696,
  4.515852212905884,
  4.595580339431763,
  4.453830718994141)}

In [ ]:
# Train the algorithm on the trainset, and predict ratings for the testset
preds_ub = algo_ub.fit(trainset).test(testset)

# RMSE to check accuracy
accuracy.rmse(preds_ub)

Estimating biases using als...
Computing the cosine similarity matrix...
Done computing similarity matrix.
RMSE: 0.9258


0.9257665505762505

## Results
**User-based collaborative filtering**:

    - Trainset 5-fold Cross validation RMSE = 0.9340 ± 0.0044
    - Testset RMSE = 0.9258

## Item-based collaborative Filtering
- KNN algorithm with top 30 neighbours/items for prediction
- similarity measure : cosine similarity
- Stochastic Gradient Descent(SGD) for baseline estimation
- Cross-validating the approach on the training set
- using RMSE on the testing set for accuracy

In [ ]:
sim_options = {"name": "cosine", "user-based": False}
algo_ib = KNNBaseline(k=40, sim_options=sim_options)

In [ ]:
# Run 5-fold cross-validation
cross_validate(algo_ib, data, measures=['RMSE'], cv=5, verbose=True)

Estimating biases using als...
Computing the cosine similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the cosine similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the cosine similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the cosine similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the cosine similarity matrix...
Done computing similarity matrix.
Evaluating RMSE of algorithm KNNBaseline on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9286  0.9388  0.9315  0.9441  0.9240  0.9334  0.0072  
Fit time          1.47    1.56    1.56    1.55    1.52    1.53    0.03    
Test time         4.62    4.75    4.66    4.79    4.58    4.68    0.08    


{'fit_time': (1.4673869609832764,
  1.5580594539642334,
  1.5581674575805664,
  1.5476727485656738,
  1.5209035873413086),
 'test_rmse': array([0.92858902, 0.93878331, 0.9315013 , 0.94413781, 0.92396077]),
 'test_time': (4.6206138134002686,
  4.753770589828491,
  4.662949085235596,
  4.79037618637085,
  4.583998203277588)}

In [ ]:
# Train the algorithm on the trainset, and predict ratings for the testset
preds_ib = algo_ib.fit(trainset).test(testset)

# RMSE to check accuracy
accuracy.rmse(preds_ib)

Estimating biases using als...
Computing the cosine similarity matrix...
Done computing similarity matrix.
RMSE: 0.9258


0.9257665505762505

## Results
**Item-based collaborative filtering**:

    - Trainset 5-fold Cross validation RMSE = 0.9334 ± 0.0072
    - Testset RMSE = 0.9343